File: internal/cloudauth/pairing.go
package cloudauth

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"fmt"
	"io"
	"strings"

	"golang.org/x/crypto/chacha20poly1305"
	"golang.org/x/crypto/curve25519"
	"golang.org/x/crypto/hkdf"
)

// Relay-blind pairing via ephemeral X25519 ECDH (docs/specs/agentworks-wire.md §7,
// cloud/crypto/src/pairing.ts). Both sides derive the same key from the ECDH
// shared secret via HKDF-SHA256 with info "linx-pair-v1" (FROZEN v1 — must match
// the TypeScript reference byte-for-byte) and wrap the master secret with
// XChaCha20-Poly1305. Wire ciphertext = base64url(nonce(24) || ct).
const pairInfo = "linx-pair-v1"

// pairingKey derives the symmetric pairing key from an X25519 shared secret.
func pairingKey(shared []byte) ([]byte, error) {
	r := hkdf.New(sha256.New, shared, nil, []byte(pairInfo))
	key := make([]byte, 32)
	if _, err := io.ReadFull(r, key); err != nil {
		return nil, err
	}
	return key, nil
}

// decodePairB64url decodes the wire ciphertext. @scure/base `base64url` and Go
// agree byte-for-byte for a wrapped 32-byte secret (72 bytes → no padding); we
// strip any padding to be tolerant either way.
func decodePairB64url(s string) ([]byte, error) {
	return base64.RawURLEncoding.DecodeString(strings.TrimRight(s, "="))
}

// NewPairingEphemeral generates an X25519 ephemeral keypair, returning the public
// key as lowercase hex plus the raw 32-byte private key.
func NewPairingEphemeral() (pubHex string, priv []byte, err error) {
	priv = make([]byte, 32)
	if _, err = rand.Read(priv); err != nil {
		return "", nil, err
	}
	pub, err := curve25519.X25519(priv, curve25519.Basepoint)
	if err != nil {
		return "", nil, err
	}
	return hex.EncodeToString(pub), priv, nil
}

// PairingComplete is the receiver side (a secret-less machine): derive the shared
// key from the holder's ephemeral public key and decrypt the wrapped master
// secret. Errors on a bad key/tamper (AEAD tag) or malformed input.
func PairingComplete(priv []byte, holderPubHex, ciphertextB64url string) ([]byte, error) {
	holderPub, err := hex.DecodeString(holderPubHex)
	if err != nil {
		return nil, fmt.Errorf("decode holder pubkey: %w", err)
	}
	shared, err := curve25519.X25519(priv, holderPub)
	if err != nil {
		return nil, fmt.Errorf("x25519: %w", err)
	}
	key, err := pairingKey(shared)
	if err != nil {
		return nil, err
	}
	raw, err := decodePairB64url(ciphertextB64url)
	if err != nil {
		return nil, fmt.Errorf("decode ciphertext: %w", err)
	}
	if len(raw) < chacha20poly1305.NonceSizeX {
		return nil, fmt.Errorf("pairing ciphertext too short")
	}
	aead, err := chacha20poly1305.NewX(key)
	if err != nil {
		return nil, err
	}
	pt, err := aead.Open(nil, raw[:chacha20poly1305.NonceSizeX], raw[chacha20poly1305.NonceSizeX:], nil)
	if err != nil {
		return nil, fmt.Errorf("pairing decrypt failed (tamper/wrong key): %w", err)
	}
	return pt, nil
}

// PairingRespond is the holder side (has the secret): generate an ephemeral
// keypair, derive the shared key from the receiver's ephemeral public key, and
// wrap masterSecret. Returns the holder's ephemeral public key (hex) and
// base64url(nonce(24) || ct). Primarily for tests and CLI symmetry — in the
// unified system the authenticated web/linx performs this step.
func PairingRespond(masterSecret []byte, receiverPubHex string) (pubHex, ciphertextB64url string, err error) {
	receiverPub, err := hex.DecodeString(receiverPubHex)
	if err != nil {
		return "", "", fmt.Errorf("decode receiver pubkey: %w", err)
	}
	priv := make([]byte, 32)
	if _, err = rand.Read(priv); err != nil {
		return "", "", err
	}
	pub, err := curve25519.X25519(priv, curve25519.Basepoint)
	if err != nil {
		return "", "", err
	}
	shared, err := curve25519.X25519(priv, receiverPub)
	if err != nil {
		return "", "", fmt.Errorf("x25519: %w", err)
	}
	key, err := pairingKey(shared)
	if err != nil {
		return "", "", err
	}
	nonce := make([]byte, chacha20poly1305.NonceSizeX)
	if _, err = rand.Read(nonce); err != nil {
		return "", "", err
	}
	aead, err := chacha20poly1305.NewX(key)
	if err != nil {
		return "", "", err
	}
	ct := aead.Seal(nil, nonce, masterSecret, nil)
	wire := append(append([]byte{}, nonce...), ct...)
	return hex.EncodeToString(pub), base64.RawURLEncoding.EncodeToString(wire), nil
}

File: cloud/crypto/src/pairing.ts
/**
 * Relay-blind pairing via ephemeral X25519 ECDH.
 *
 * The web and the machine (linxd) each generate an ephemeral X25519 keypair and
 * exchange only public keys + ciphertext through the relay. They derive the same
 * symmetric key from the ECDH shared secret (HKDF-SHA256, info `linx-pair-v1`)
 * and use it to wrap the account master secret with xchacha20poly1305. The relay
 * couriers public keys and ciphertext but never learns the master secret.
 */
import { x25519 } from '@noble/curves/ed25519.js'
import { xchacha20poly1305 } from '@noble/ciphers/chacha'
import { randomBytes } from '@noble/ciphers/webcrypto'
import { hkdf } from '@noble/hashes/hkdf.js'
import { sha256 } from '@noble/hashes/sha2.js'
import { bytesToHex, hexToBytes, utf8ToBytes } from '@noble/hashes/utils.js'
import { base64url } from '@scure/base'

const KEY_BYTES = 32
const NONCE_BYTES = 24
const PAIR_INFO = utf8ToBytes('linx-pair-v1')

/** Derive the symmetric pairing key from an X25519 shared secret. */
function pairingKey(shared: Uint8Array): Uint8Array {
  return hkdf(sha256, shared, undefined, PAIR_INFO, KEY_BYTES)
}

export interface PairingInitiator {
  /** Hex of the web side's ephemeral X25519 public key. */
  epkWebPub: string
  /**
   * Finish pairing on the web side: derive the shared key from the machine's
   * ephemeral public key, then decrypt `base64url(nonce(24) || ct)` into the
   * master secret. Throws on auth failure (tampering / wrong key).
   */
  complete(epkMachinePub: string, ciphertext: string): Uint8Array
}

/**
 * WEB side. Generate an ephemeral X25519 keypair and return its public key plus
 * a `complete()` that decrypts the master secret the machine wrapped for us.
 */
export function createPairingInitiator(): PairingInitiator {
  const webPriv = x25519.utils.randomSecretKey()
  const epkWebPub = bytesToHex(x25519.getPublicKey(webPriv))

  return {
    epkWebPub,
    complete(epkMachinePub: string, ciphertext: string): Uint8Array {
      const shared = x25519.getSharedSecret(webPriv, hexToBytes(epkMachinePub))
      const key = pairingKey(shared)
      const raw = base64url.decode(ciphertext)
      const nonce = raw.slice(0, NONCE_BYTES)
      const ct = raw.slice(NONCE_BYTES)
      return xchacha20poly1305(key, nonce).decrypt(ct)
    },
  }
}

export interface PairingResponse {
  /** Hex of the machine side's ephemeral X25519 public key. */
  epkMachinePub: string
  /** `base64url(nonce(24) || ct)` of the wrapped master secret. */
  ciphertext: string
}

/**
 * MACHINE / linxd side. Generate an ephemeral X25519 keypair, derive the shared
 * key from the web's ephemeral public key, and wrap `masterSecret` for the web.
 */
export function pairingRespond(masterSecret: Uint8Array, epkWebPub: string): PairingResponse {
  const machinePriv = x25519.utils.randomSecretKey()
  const epkMachinePub = bytesToHex(x25519.getPublicKey(machinePriv))
  const shared = x25519.getSharedSecret(machinePriv, hexToBytes(epkWebPub))
  const key = pairingKey(shared)
  const nonce = randomBytes(NONCE_BYTES)
  const ct = xchacha20poly1305(key, nonce).encrypt(masterSecret)
  const raw = new Uint8Array(nonce.length + ct.length)
  raw.set(nonce, 0)
  raw.set(ct, nonce.length)
  return { epkMachinePub, ciphertext: base64url.encode(raw) }
}
